src/puya/awst_build/eb/arc4/abi_call.py

```diff
@@ -0,0 +1,407 @@
+        arg_kinds: list[mypy.nodes.ArgKind],
+        arg_names: list[str | None],
+        location: SourceLocation,
+        original_expr: mypy.nodes.CallExpr,
+    ) -> ExpressionBuilder:
+        abi_call_expr = _extract_abi_call_args(args, arg_kinds, arg_names, location)
+        method = abi_call_expr.method
+
+        match method:
+            case Literal(value=str() as method_str):
+                pass
+            case _:
+                raise CodeError(
+                    "First argument must be a `str` value of an ARC4 method name/selector",
+                    location,
+                )
+
+        method_name, *_ = method_str.split("(", maxsplit=1)
+        arg_wtypes = list(map(_arg_to_wtype, abi_call_expr.abi_args))
+        calculated_signature = get_abi_signature_from_wtypes(method_name, arg_wtypes, self.wtype)
+        if not calculated_signature.startswith(method_str):
+            raise CodeError(
+                f"Method selector from args '{calculated_signature}' "
+                f"does not match provided method selector: '{method_str}'",
+                method.source_location,
+            )
+
+        return var_expression(
+            _create_abi_call_expr(
+                calculated_signature,
+                arg_wtypes,
+                self.wtype,
+                abi_call_expr.abi_args,
+                abi_call_expr.transaction_kwargs,
+                location,
+            )
+        )
+
+
+def _arg_to_wtype(arg: ExpressionBuilder | Literal) -> wtypes.WType:
+    match arg:
+        case ExpressionBuilder(value_type=wtypes.WType() as wtype):
+            return wtype
+        case Literal(value=bytes()):
+            return wtypes.bytes_wtype
+        case Literal(value=int()):
+            # TODO: what about mapping int to BigUInt?
+            return wtypes.uint64_wtype
+        case Literal(value=str()):
+            return wtypes.arc4_string_wtype
+        case Literal(value=bool()):
+            return wtypes.bool_wtype
+        case _:
+            raise CodeError("Invalid arg type", arg.source_location)
+
+
+def _create_abi_call_expr(
+    method_selector: str,
+    abi_arg_types: Sequence[wtypes.WType],
+    abi_return_type: wtypes.WType,
+    abi_args: Sequence[ExpressionBuilder | Literal],
+    transaction_kwargs: dict[str, ExpressionBuilder | Literal],
+    location: SourceLocation,
+) -> Expression:
+    abi_arg_exprs: list[Expression] = [
+        MethodConstant(
+            value=method_selector,
+            source_location=location,
+        )
+    ]
+    asset_exprs = list[Expression]()
+    account_exprs = list[Expression]()
+    application_exprs = list[Expression]()
+
+    def append_ref_arg(ref_list: list[Expression], arg_expr: Expression) -> None:
+        # asset refs start at 0, account and application start at 1
+        implicit_offset = 0 if ref_list is asset_exprs else 1
+        # TODO: what about references that are used more than once?
```

Comment:
🤔

Reply:
The off chain SDK's handle this by checking what has already been added and only adding new references, however doing that chain is a bit cost prohibitive and not sure if its neccessary

Reply:
For now leaving this as a TODO, will address in the future